corfudb-tools/src/main/java/org/corfudb/compactor/CorfuStoreCompactorMain.java

```diff
@@ -0,0 +1,327 @@
+        }
+        try (TxnContext txn = corfuStore.txn(CORFU_SYSTEM_NAMESPACE)) {
+            txn.putRecord(checkpointTable, DistributedCompactor.UPGRADE_KEY, TokenMsg.getDefaultInstance(),
+                    null);
+            txn.commit();
+        } catch (Exception e) {
+            log.warn("Unable to write UpgradeKey to checkpoint table, ", e);
+        }
+    }
+
+    /**
+     * Entry point to invoke Client checkpointing by the CorfuServer
+     *
+     * @param args command line argument strings
+     */
+    @SuppressWarnings("UncommentedMain")
+    public static void main(String[] args) throws Exception {
+        getCompactorArgs(args);
+
+        if (maxWriteSize == -1) {
+            if ("".equals(persistedCacheRoot)) {
+                // in-memory compaction
+                maxWriteSize = DEFAULT_CP_MAX_WRITE_SIZE;
+                Thread.currentThread().setName("CS-Config-chkpter");
+            } else {
+                // disk-backed non-config compaction
+                maxWriteSize = NON_CONFIG_DEFAULT_CP_MAX_WRITE_SIZE;
+                Thread.currentThread().setName("CS-NonConfig-chkpter");
+            }
+        }
+
+        CorfuStoreCompactorMain corfuCompactorMain = new CorfuStoreCompactorMain();
+
+        if (isUpgrade) {
+            corfuCompactorMain.upgrade();
+        }
+        corfuCompactorMain.checkpoint();
+    }
+
+    private void checkpoint() {
+        try {
+            for (int i = 0; i< retryCheckpointing; i++) {
+                if (distributedCompactor.startCheckpointing() > 0) {
+                    break;
+                }
+                TimeUnit.SECONDS.sleep(1);
+            }
+        } catch (Throwable throwable) {
+            log.error("CorfuStoreCompactorMain crashed with error:", CORFU_LOG_CHECKPOINT_ERROR, throwable);
+        }
+    }
+
+    /**
+     * Create a protobuf serializer.
+     *
+     * @return Protobuf Serializer.
+     */
+    private ISerializer createProtobufSerializer() {
+        ConcurrentMap<String, Class<? extends Message>> classMap = new ConcurrentHashMap<>();
+
+        // Register the schemas of TableName, TableDescriptors, TableMetadata, ProtobufFilename/Descriptor
+        // to be able to understand registry table.
+        classMap.put(getTypeUrl(TableName.getDescriptor()), TableName.class);
+        classMap.put(getTypeUrl(TableDescriptors.getDescriptor()),
+                TableDescriptors.class);
+        classMap.put(getTypeUrl(TableMetadata.getDescriptor()),
+                TableMetadata.class);
+        classMap.put(getTypeUrl(ProtobufFileName.getDescriptor()),
+                ProtobufFileName.class);
+        classMap.put(getTypeUrl(ProtobufFileDescriptor.getDescriptor()),
+                ProtobufFileDescriptor.class);
+        return new ProtobufSerializer(classMap);
+    }
+
+    /**
+     * Populate the ProtobufDescriptorTable using the RegistryTable.
+     * Enables backward compatibility in case of data migration.
```

Comment:
does the registry table get modified during data migration?  Can you explain when the protobufDescriptor table needs to be synced?

Reply:
in case we are upgrading from corfu-0.3.0 branch where there was no protobuf descriptor table. In those cases the protobuf descriptor table needs to be constructed or synced from the table registry. This code is just copying that logic from the old compactor